### DISCUSSION TITLE
v1.5.6 how to add XYLineAnnotation?

### DISCUSSION THREAD

**Discussion by matoos32** [matoos32/jfreechart-builder#42] [Ideas]
In response to https://github.com/jfree/jfreechart/issues/197#issuecomment-1304795558

@jcobb88 wrote: 
> @matoos32 Like your library very much. Your API allows to add simply annotations, but how do you add something like XYLineAnnotation, which require two x and y values? Adding those directly to the XYPlot obviously doesn't work when we disable gaps. If not possible, what would be a workaround?

@jcobb88 thanks for your kind feedback. Happy it's been useful.

I examined the [XYAnnotation](https://github.com/jfree/jfreechart/blob/v1.5.3/src/main/java/org/jfree/chart/annotations/XYAnnotation.java) type hierarchy and I see each implementing class has a different constructor. [IXYAnnotationBuilder](https://github.com/matoos32/jfreechart-builder/blob/v1.5.6/src/main/java/com/jfcbuilder/builders/IXYAnnotationBuilder.java) was quickly coded around [XYPointerAnnotation](https://github.com/jfree/jfreechart/blob/v1.5.3/src/main/java/org/jfree/chart/annotations/XYPointerAnnotation.java) without assessing all of the jfreechart annotation classes. So [IXYAnnotationBuilder](https://github.com/matoos32/jfreechart-builder/blob/v1.5.6/src/main/java/com/jfcbuilder/builders/IXYAnnotationBuilder.java) will likely need additional methods to suit all the XYAnnotation implementations, but that doesn't scale well and will get ugly. So it needs some thought. I'll create a project to support all annotation types.

In the mean time I found you can add the XYLineAnnotation to the disabled-gap chart using the integer indices of desired date-axis points. See the snippet below to further decorate the built chart.

Inserting new charts in the demo project [JFreeChartBuilderDemo](https://github.com/matoos32/jfreechart-builder-demo/blob/v1.5.6/src/main/java/com/jfcbuilder/demo/JFreeChartBuilderDemo.java#L319):

``` 
    JFreeChart chart = getDailyStockChartBuilder().title("XYLineAnnotation Example With Time Gaps").build();
    CombinedDomainXYPlot combinedPlot = (CombinedDomainXYPlot) chart.getPlot();
    @SuppressWarnings("unchecked")
    List<XYPlot> subplots = combinedPlot.getSubplots();
    XYPlot firstPlot = subplots.stream().findFirst().get();
    long[] timeArray = dohlcv.dates();
    final int index1 = timeArray.length - 10;
    final int index2 = index1 + 5;
    final long date1 = timeArray[index1];
    final long date2 = timeArray[index2];
    final double price1 = dohlcv.highs()[index1];
    final double price2 = dohlcv.highs()[index2];
    firstPlot.addAnnotation(new XYLineAnnotation(date1, price1, date2, price2, BuilderConstants.DASHED_LINE, Color.MAGENTA));
    
    charts.add(chart);
    
    JFreeChart chart2 = getDailyStockChartBuilder().title("XYLineAnnotation Example NO Time Gaps").showTimeGaps(false).build();
    CombinedDomainXYPlot combinedPlot2 = (CombinedDomainXYPlot) chart2.getPlot();
    @SuppressWarnings("unchecked")
    List<XYPlot> subplots2 = combinedPlot2.getSubplots();
    firstPlot = subplots2.stream().findFirst().get();
    final int plot2Index1 = timeArray.length - ohlcStartIndex - 10;
    final int plot2Index2 = plot2Index1 + 5;
    firstPlot.addAnnotation(new XYLineAnnotation(plot2Index1, price1, plot2Index2, price2, BuilderConstants.DASHED_LINE, Color.MAGENTA));
    
    charts.add(chart2);
```

**Comment by matoos32** [comment]
Now planned and tracked in this project: https://github.com/users/matoos32/projects/1

**Comment by matoos32** [comment]
@jcobb88 I added a builder for XYLineAnnotation.

It's in the develop branch as of pull request https://github.com/matoos32/jfreechart-builder/pull/58.

There's an example in the demo app I just migrated in ([JFreeChartBuilderDemo.java#L304](https://github.com/matoos32/jfreechart-builder/blob/develop/demo-app/src/main/java/com/jfcbuilder/demo/JFreeChartBuilderDemo.java#L304))

Give it a go and let me know what you think!

**Reply by jcobb88** [reply]
@matoos32 Looks good, thanks! As you know I suggested also some other ideas. 
I did a lot of workarounds by getting the specific XYPlots and replacing/adding the annotations. So that works quite well. Some things are a little less clear, e.g. changing the tick labels, so if your time is restricted you may focus on those things and add the workaround for changing data on the plot: retrieving subplots and adding annotations directly (using indices for gapless charts).

**Reply by matoos32** [reply]
@jcobb88 I can look at tick labels next. https://github.com/matoos32/jfreechart-builder/discussions/53

I'm unclear if you still have an issue with gapless plot annotations not drawn correctly when updating data. If so, could you submit a bug issue showing how to reproduce it?
